### PR TITLE
feat: add app initializer

### PR DESCRIPTION
**Pull request by supproduction**
adds useExisting provider
adds app initializer service
adds initializers token

closes [HRZ-2541](https://spryker.atlassian.net/browse/HRZ-2541)

[HRZ-2541]: https://spryker.atlassian.net/browse/HRZ-2541?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: libs/platform/core/src/services/app-initializer/app-initializer.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+export const AppInitializerService = 'oryx.AppInitializerService';
+export const AppInitializer = 'oryx.AppInitializer*';
+
+export interface AppInitializerService extends OnDestroy {
+  initialize(): void;
+}
+
+export interface AppInitializer {
+  initialize(): void | Observable<unknown>;
```
Comment (gund) [inline libs/platform/core/src/services/app-initializer/app-initializer.ts]: Return type should be consistent, also maybe we can allow for sync methods as well:
```suggestion
  initialize(): void | Promise<void> | Observable<void>;
```

File: libs/platform/core/src/services/app-initializer/app-initializer.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import { OnDestroy } from '@spryker-oryx/di';
+import { Observable } from 'rxjs';
+
+export const AppInitializerService = 'oryx.AppInitializerService';
+export const AppInitializer = 'oryx.AppInitializer*';
+
+export interface AppInitializerService extends OnDestroy {
```
Comment (gund) [inline libs/platform/core/src/services/app-initializer/app-initializer.ts]: Why do you need to enforce that `AppInitializerService` should also implement `OnDestroy`?
Shouldn't this be implementation details?

**Review by gund — COMMENTED** [review]
Looks good just a few small suggestions.

**Review by gund — APPROVED** [review]

File: libs/platform/core/src/services/app-initializer/app-initializer.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
+import { Observable } from 'rxjs';
+
+export const AppInitializerService = 'oryx.AppInitializerService';
+export const AppInitializer = 'oryx.AppInitializer*';
+
+export interface AppInitializerService {
+  initialize(): Promise<void>;
+}
+
+export interface AppInitializer {
```
Comment (dunqan) [inline libs/platform/core/src/services/app-initializer/app-initializer.ts:10]: would be nice to be able to use just promise or observable, not full blown class

File: libs/platform/core/src/services/app-initializer/default-app-initializer.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+  async initialize(): Promise<void> {
+    const initializers = this.injector.inject(AppInitializer, []);
+
+    for (const initializer of initializers) {
+      const result = initializer.initialize();
+
+      if (isPromise(result)) {
+        await result;
+      }
```
Comment (dunqan) [inline libs/platform/core/src/services/app-initializer/default-app-initializer.ts:17]: shouldn't we use `else` here?

**Pull request merged**